player/lua/select.lua

```diff
@@ -353,6 +353,118 @@ mp.add_key_binding(nil, "select-audio-device", function ()
     })
 end)
 
+local history_file_path =
+    mp.command_native({"expand-path", mp.get_property("watch-history-path")})
+
+mp.register_event("file-loaded", function ()
+    if not mp.get_property_native("save-watch-history") then
+        return
+    end
+
+    local history_file, error_message = io.open(history_file_path, "a")
+    if not history_file then
+        show_error("Failed to write the watch history: " .. error_message)
+        return
+    end
+
+    local path = mp.command_native({"normalize-path", mp.get_property("path")})
+    local title = mp.get_property("playlist/" .. mp.get_property("playlist-pos") .. "/title")
+
+    history_file:write(utils.format_json({os.time(), path, title}) .. "\n")
+    history_file:close()
+end)
+
+local function get_history_entry(line, seen)
+    local entry = utils.parse_json(line)
+
+    if not entry then
+        mp.msg.warn(line .. " in " .. history_file_path .. " is not valid JSON.")
+        return
+    end
+
+    local time, path, title = unpack(entry)
+
+    if not path then
+        mp.msg.warn(line .. " in " .. history_file_path .. " has invalid data.")
+        return
+    end
+
+    if seen[path] then
+        return
+    end
+
+    seen[path] = true
+
+    local status, date = pcall(os.date, "(%Y-%m-%d %H:%M) ", time)
+
+    if not status then
+        mp.msg.warn(line .. " in " .. history_file_path .. " has invalid data.")
+        return
+    end
+
+    return {
+        ["date"] = date,
+        ["path"] = path,
+        ["title"] = title,
+    }
+end
+
+mp.add_key_binding(nil, "select-watch-history", function ()
+    local history_file, error_message = io.open(history_file_path)
+    if not history_file then
+        show_warning(mp.get_property_native("save-watch-history")
+                     and error_message
+                     or "Enable --save-watch-history")
+        return
+    end
+
+    local lines = {}
+    for line in history_file:lines() do
+        table.insert(lines, line)
+    end
+    history_file:close()
+
+    local entries = {}
+    local seen = {}
+    for i = #lines, 1, -1 do
+        local entry = get_history_entry(lines[i], seen)
+        if entry then
+            entries[#entries + 1] = entry
+        end
+    end
+
+    local items = {}
+    local osd_playlist_entry = mp.get_property("osd-playlist-entry")
+    for i, entry in pairs(entries) do
+        local item = entry.path
+        if entry.title and osd_playlist_entry == "title" then
+            item = entry.title
+        elseif entry.title and osd_playlist_entry == "both" then
+            item = entry.title .. " (" .. entry.path .. ")"
+        end
+        items[i] = entry.date .. item
+    end
+    items[#items+1] = "Clear history"
```

Comment:
```suggestion
    items[#items+1] = "Clear history"
    entries = nil
    seen = nil
```

Just to make GC know we don't need them anymore.

Reply:
`entries` is used in `submit`. Cleared  `lines` and `seen` together above.

Reply:
Made the linter fail.

player/lua/select.lua:435:5: value assigned to variable lines is unused
player/lua/select.lua:436:5: value assigned to variable seen is unused

Reply:
pitty